### PR TITLE
hostapd: add ability to allow client after certain connection tries

### PR DESCRIPTION
**Pull request by PolynomialDivision**
If you want to use del_client for load balancing there should be some mechanism, to allow a client to still connect to the AP although it is banned. So the ban should be removed after a number of connection tries.

Signed-off-by: Nick Hainke <vincent@systemli.org>

Update:
Should work now. :)

### CODE REVIEW

**Comment by ynezz** [comment]
> If you want to use del_client for load balancing 

It just seems to me, that you're bending/abusing/misusing `del_client` command for some other purpose. Maybe you need to think about it again and come up with better and more clear solution? Or am I missing something?

>there should be some mechanism, to allow a client to still connect to the AP although it is banned. 

If the client is banned, then it's banned. I don't understand why it should be allowed to connect until the  ban timeout has expired.

> So the ban should be removed after a number of connection tries.

I don't think so. The ban is going to be removed after the ban timeout.

Since nobody has touched this for about 6 months, it's very unlikely, that it's going to be merged as it is.

**Comment by PolynomialDivision** [comment]
> > If you want to use del_client for load balancing
> 
> It just seems to me, that you're bending/abusing/misusing `del_client` command for some other purpose. Maybe you need to think about it again and come up with better and more clear solution? Or am I missing something?

I don't think that I'm the only one who misused this call to do client steering,load balancing. I know that there is 802.11v... I did not know that there is some policy for this call.

> > there should be some mechanism, to allow a client to still connect to the AP although it is banned.
> 
> If the client is banned, then it's banned. I don't understand why it should be allowed to connect until the ban timeout has expired.

If you ban a client and hope it will connect to another AP in his range, you will notice that this can lead to unexpected behvior. Some driver implementation do not care about status codes or disassoc/deauth reasons, and just try to connect and connect again to the same AP. If you want to deal with that behavior in an easy way, you can remove this ban if the client connects several times.
> > So the ban should be removed after a number of connection tries.
> 
> I don't think so. The ban is going to be removed after the ban timeout.

This is still possible (backwards compatible)... If you do not include the `count parameters` in the ubus call it will work like it did in the past.
> Since nobody has touched this for about 6 months, it's very unlikely, that it's going to be merged as it is.
It always needs some time that a person is merging the stuff in their staging tree. I don't rely on this patch, but I think it is a nice feature.

**Comment by CodeFetch** [comment]
@PolynomialDivision The name "allow_con_count" does not make clear what it does. E.g. "ban_for_max_attempts" or "max_ban_attempts" makes it a bit more clear, but it might still be confusing. Thus you should also add some comments. Hostapd is sparely commented. It's a shame.

**Comment by PolynomialDivision** [comment]
> @PolynomialDivision The name "allow_con_count" does not make clear what it does. E.g. "ban_for_max_attempts" or "max_ban_attempts" makes it a bit more clear, but it might still be confusing. Thus you should also add some comments. Hostapd is sparely commented. It's a shame.

I can change that, but is this patch still discussed/wanted? Maybe I should just close and reopen, if I need that in the future.

**Comment by CodeFetch** [comment]
@sargon is surely happy about this as he has a [tool](https://github.com/sargon/roamguide) that might use this feature.

**Comment by PolynomialDivision** [comment]
> @sargon is surely happy about this as he has a [tool](https://github.com/sargon/roamguide) that might use this feature.

Okay. I will rewrite it. ;)

**Comment by PolynomialDivision** [comment]
> The name "allow_con_count" does not make clear what it does. E.g. "ban_for_max_attempts" or "max_ban_attempts" makes it a bit more clear

Renamed. :)

**Comment by CodeFetch** [comment]
@ynezz As PolynomialDivision stated the behavior does not change unless you set the parameter which makes it backwards-compatible. 

This feature is e.g. useful for https://github.com/sargon/roamguide
Cisco has a similar feature called "Optimized Roaming".
The problems with 802.11v are that it must be supported by the client and the AP needs to provide a candidate list with alternative BSSIDs to use for the STA in the BSS transition request, but there are scenarios where the AP does not know the BSSIDs of other APs the STA could choose.

**Comment by ynezz** [comment]
@CodeFetch Ok, so can I have your Tested-by ? Please also state which devices/configs you've actually tested it. Thanks.

File: package/network/services/hostapd/src/src/ap/ubus.c [hunk 1]
```diff
@@ -130,6 +132,8 @@ hostapd_bss_ban_client(struct hostapd_data *hapd, u8 *addr, int time)
 		ban = os_zalloc(sizeof(*ban));
 		memcpy(ban->addr, addr, sizeof(ban->addr));
 		ban->avl.key = ban->addr;
+		ban->max_ban_attempts = max_ban_attempts;
+		ban->con_tries = 0;
```
Comment (ynezz) [inline package/network/services/hostapd/src/src/ap/ubus.c]: this is superfluous as it's zeroed already in os_zalloc

File: package/network/services/hostapd/src/src/ap/ubus.c [hunk 2]
```diff
@@ -269,6 +275,7 @@ hostapd_bss_del_client(struct ubus_context *ctx, struct ubus_object *obj,
 	struct sta_info *sta;
 	bool deauth = false;
 	int reason;
+	int max_ban_attempts;
```
Comment (ynezz) [inline package/network/services/hostapd/src/src/ap/ubus.c]: Can't you just set it to 0 here so you don't need that else block bellow (there is a stray white space after that else anyway)?

**Comment by CodeFetch** [comment]
Okay, I'm going to make some tests.

File: package/network/services/hostapd/src/src/ap/ubus.c [hunk 3]
```diff
@@ -1053,7 +1065,18 @@ int hostapd_ubus_handle_event(struct hostapd_data *hapd, struct hostapd_ubus_req
 
 	ban = avl_find_element(&hapd->ubus.banned, addr, ban, avl);
 	if (ban)
-		return WLAN_STATUS_AP_UNABLE_TO_HANDLE_NEW_STA;
+	{
```
Comment (ynezz) [inline package/network/services/hostapd/src/src/ap/ubus.c]:
```
if (ban) {
} else {
}
```

Please try to follow the coding style already established in that file.

File: package/network/services/hostapd/src/src/ap/ubus.c [hunk 4]
```diff
@@ -1053,7 +1065,18 @@ int hostapd_ubus_handle_event(struct hostapd_data *hapd, struct hostapd_ubus_req
...
+		if(0 < ban->max_ban_attempts && ban->con_tries >= ban->max_ban_attempts)
```
Comment (ynezz) [inline package/network/services/hostapd/src/src/ap/ubus.c]: missing space after if so it should be `if () {`

Comment (ynezz) [reply package/network/services/hostapd/src/src/ap/ubus.c]: That logic seems to be weird and quite hairy as well. You should probably make `max_ban_attempts` unsigned int and test here for `if (ban->max_ban_attempts > 0 && ban->con_tries >= ban->max_ban_attempts)` which probably means, that con_tries should be unsigned int as well.

File: package/network/services/hostapd/src/src/ap/ubus.c [hunk 5]
```diff
@@ -285,6 +292,11 @@ hostapd_bss_del_client(struct ubus_context *ctx, struct ubus_object *obj,
 	if (tb[DEL_CLIENT_DEAUTH])
 		deauth = blobmsg_get_bool(tb[DEL_CLIENT_DEAUTH]);
 
+	if (tb[DEL_CLIENT_ALLOW_CON_COUNT])
+		max_ban_attempts = blobmsg_get_u32(tb[DEL_CLIENT_ALLOW_CON_COUNT]);
```
Comment (ynezz) [inline package/network/services/hostapd/src/src/ap/ubus.c]: Wouldn't it make more sense to rename `DEL_CLIENT_ALLOW_CON_COUNT` into `DEL_CLIENT_BAN_MAX_ATTEMPTS` (or rename the max_ban_attempts to something else)? It doesn't make sense to name it allow_con_count and assign it to max_ban_attempts variable, I really miss the point here.

**Comment by PolynomialDivision** [comment]
@ynezz Thanks for your remarks. :)
I changed the code accordingly. 

@CodeFetch Can you test it? Right now I do not have an AP to test.

At the end, all this denying of authentication/association requests and de-association LB approaches are a bit hairy... That is why I want to try different things. On my fork, I start implementing `csa-handovers`... https://github.com/PolynomialDivision/openwrt/commit/12e8bfaa15f0b4b769a8986b1d4c852c38c8a654

**Comment by CodeFetch** [comment]
@PolynomialDivision Hm... I don't really understand how that helps with helping STAs to roam, but thank you for that. It's very nice for mesh networks. It would be nice if you could implement the 802.11v using ubus later. This way we could do proper roam-guidance (if STA supports 802.11v and we have a list of alternative BSSIDs configured, use that else use the ban approach).

**Comment by PolynomialDivision** [comment]
You can read this paper from cisco people: [ClientMarshal](https://www.researchgate.net/publication/329933591_ClientMarshal_Regaining_Control_from_Wireless_Clients_for_Better_Experience_in_Enterprise_WLANs) Using CSA for seamless handover with legacy clients.

**Comment by ynezz** [comment]
> Okay, I'm going to make some tests.

@CodeFetch so? :-)

**Comment by CodeFetch** [comment]
@ynezz I've asked @sargon to make some tests, because I'm very busy at the moment and he knows better how it should behave. I'll do it if he won't.

**Comment by sargon** [comment]
I am kind of busy myself, don't expect any results before camp

**Comment by PolynomialDivision** [comment]
I started to go different ways, using `802.11v`. I will close.

**Comment by CodeFetch** [comment]
@PolynomialDivision ClientMarshal in OpenWrt :flushed:? If so please drop a link to your development tree and reopen or open a new pull-request with a title like "Client steering/802.11v" or something to make other people aware.

**Comment by PolynomialDivision** [comment]
@CodeFetch
Nope, I programm DAWN: https://github.com/berlin-open-wireless-lab/DAWN
Still buggy and fixing bugs.

Here are some forum posts:
- https://forum.openwrt.org/t/finally-got-802-11v-working-fully-few-pointers/56678/58
- https://forum.openwrt.org/t/how-does-rrm-work/32635/61
- https://github.com/berlin-open-wireless-lab/DAWN/issues/60

Just try install dawn with opkg and the luci-app-dawn. Feel free to open an issue. Currently, I do not have mich time to go further... :/ Still need to figure out how to compare rssi, rcpi and rsni. And I think my bss transition call does not work with every station. Still have to handle legacy clients... Let's see... :)

**Comment by CodeFetch** [comment]
@PolynomialDivision Wow! That's something I tried to acquire funding for. Thus I'm very happy about BOWL taking care of it.

> Still need to figure out how to compare rssi, rcpi and rsni (802.11k).

I tried something similar for WLAN trilateration some years ago. It works for Qualcomm chipsets quite well, because they are being tuned with golden devices, but we would have needed to do some machine learning for getting proper conversions on other chipsets. Actually we didn't have a fixed conversion scheme as it is not needed for trilateration, but if you have proper measurement tools it should be possible to create calibration formulas for Qualcomm chipsets. We only looked at the beacons, because signals with higher data rates can't be used without computation intensive conversions. I think sensitivity differences on higher data rates are negligible.

**Pull request closed without merging**